_posts/2016-10-12-edsls-part-2.md

```diff
@@ -0,0 +1,413 @@
+      EitherT(F.get.map(Right(_)))
+
+    def set(s: S): EitherT[F, E, Unit] =
+      EitherT(F.set(s).map(Right(_)))
+
+    def flatMap[A, B](fa: EitherT[F, E, A])
+                     (f: A => EitherT[F, E, B]): EitherT[F, E, B] =
+      fa.flatMap(f)
+
+    def pure[A](x: A): EitherT[F, E, A] =
+      EitherT.pure(x)
+
+    def tailRecM[A, B](a: A)(f: A => EitherT[F, E, Either[A, B]]): EitherT[F, E, B] =
+      EitherT.catsDataMonadErrorForEitherT[F, E].tailRecM(a)(f)
+  }
+
+implicit def stateTMonadError[F[_], E, S](implicit F: MonadError[F, E]): MonadError[StateT[F, S, ?], E] =
+  new MonadError[StateT[F, S, ?], E] {
+    def handleErrorWith[A](fa: StateT[F, S, A])(f: E => StateT[F, S, A]): StateT[F, S, A] =
+      StateT[F, S, A] { (s: S) =>
+        val state: F[(S, A)] = fa.run(s)
+        F.handleErrorWith(state)(e => f(e).run(s))
+      }
+
+    def raiseError[A](e: E): StateT[F, S, A] =
+      StateT.lift(F.raiseError(e))
+
+    def flatMap[A, B](fa: StateT[F, S, A])(f: A => StateT[F, S, B]): StateT[F, S, B] =
+      fa.flatMap(f)
+
+    def pure[A](x: A): StateT[F, S, A] = StateT.pure(x)
+
+    def tailRecM[A, B](a: A)(f: A => StateT[F, S, Either[A, B]]): StateT[F, S, B] =
+      StateT.catsDataMonadStateForStateT[F, S].tailRecM(a)(f)
+  }
+
+type App1[A] = EitherT[State[Int, ?], String, A]
+
+type App2[A] = StateT[Either[String, ?], Int, A]
+```
+
+```scala
+val app1 = program[App1]
+// app1: App1[Int] = EitherT(cats.data.StateT@33ffe601)
+
+val app2 = program[App2]
+// app2: App2[Int] = cats.data.StateT@7f432a18
+```
+
+# Composing languages
+
+From one angle we can view our set language, or more generally any EDSL
+in this style, an effect like `MonadError` and `MonadState`. From another
+angle we can view `MonadError` and `MonadState` as EDSLs that talk about errors
+and stateful computations. We can eliminate the distinctions by renaming
+`SetLang` to `MonadSet` and treating it as a type class.
+
+```scala
+import cats.Monad
+import cats.implicits._
+
+trait MonadSet[F[_]] {
+  def monad: Monad[F]
+
+  def add(i: Int): F[Unit]
+  def remove(i: Int): F[Unit]
+  def exists(i: Int): F[Boolean]
+}
+```
+
+Composing multiple languages then becomes adding constraints to functions, and
+interpretation becomes type class instances.
+
+```scala
+trait MonadCalc[F[_]] {
+  def monad: Monad[F]
+
+  def lit(i: Int): F[Int]
+  def plus(l: F[Int], r: F[Int]): F[Int]
+}
+
+def setProgram[F[_]: MonadSet](i: Int): F[Boolean] =
+  implicitly[MonadSet[F]].exists(i)
+
+def calcProgram[F[_]: MonadCalc]: F[Int] = {
+  val calc = implicitly[MonadCalc[F]]
+  calc.plus(calc.lit(1), calc.lit(2))
+}
+
+def composedProgram[F[_]: MonadCalc: MonadSet]: F[Boolean] = {
+  implicit val monad: Monad[F] = implicitly[MonadCalc[F]].monad
+  for {
+    i <- calcProgram[F]
+    b <- setProgram(i)
+  } yield b
+}
+
+// Instance
+
+type SetState[A] = State[Set[Int], A]
+implicit val stateInstance: MonadSet[SetState] with MonadCalc[SetState] =
```

Comment:
Here, you have successfully modularised and composed programs, but not the underlying implementations. It could be worth mentioning that you can also split up this instance into two.